application_sandbox/overlapping_frames/main.cpp

```diff
@@ -0,0 +1,649 @@
+        VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO,
+        nullptr,
+        render_pass_triangle,
+        framebuffers_triangle[current_frame].get_raw_object(),
+        {{0, 0}, {app.swapchain().width(), app.swapchain().height()}},
+        1,
+        &clear_color
+    };
+
+    auto& tri_cmd_buf = command_buffers_triangle[current_frame];
+
+    app.BeginCommandBuffer(&tri_cmd_buf);
+
+    tri_cmd_buf->vkCmdBeginRenderPass(tri_cmd_buf, &pass_begin_triangle, VK_SUBPASS_CONTENTS_INLINE);
+    tri_cmd_buf->vkCmdBindPipeline(tri_cmd_buf, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_triangle);
+    tri_cmd_buf->vkCmdDraw(tri_cmd_buf, 3, 1, 0, 0);
+    tri_cmd_buf->vkCmdEndRenderPass(tri_cmd_buf);
+    LOG_ASSERT(==, data->logger(), VK_SUCCESS,
+        app.EndAndSubmitCommandBuffer(
+            &tri_cmd_buf,
+            &app.render_queue(),
+            {},
+            {},
+            {render_finished_triangle[current_frame].get_raw_object()},
+            VK_NULL_HANDLE
+        )
+    );
+
+    while(!data->WindowClosing()) {
+        // RP1
+        app.device()->vkWaitForFences(
+            app.device(),
+            1,
+            &command_trackers_post[next_frame].rendering_fence->get_raw_object(),
+            VK_TRUE,
+            UINT64_MAX
+        );
+
+        app.device()->vkResetFences(
+            app.device(),
+            1,
+            &command_trackers_post[next_frame].rendering_fence->get_raw_object()
+        );
+
+        VkRenderPassBeginInfo pass_begin_triangle {
+            VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO,
+            nullptr,
+            render_pass_triangle,
+            framebuffers_triangle[next_frame].get_raw_object(),
+            {{0, 0}, {app.swapchain().width(), app.swapchain().height()}},
+            1,
+            &clear_color
+        };
+
+        auto& tri_cmd_buf = command_buffers_triangle[next_frame];
+
+        app.BeginCommandBuffer(&tri_cmd_buf);
+
+        tri_cmd_buf->vkCmdBeginRenderPass(tri_cmd_buf, &pass_begin_triangle, VK_SUBPASS_CONTENTS_INLINE);
+        tri_cmd_buf->vkCmdBindPipeline(tri_cmd_buf, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_triangle);
+        tri_cmd_buf->vkCmdDraw(tri_cmd_buf, 3, 1, 0, 0);
+        tri_cmd_buf->vkCmdEndRenderPass(tri_cmd_buf);
+        LOG_ASSERT(==, data->logger(), VK_SUCCESS,
+            app.EndAndSubmitCommandBuffer(
+                &tri_cmd_buf,
+                &app.render_queue(),
+                {},
+                {},
+                {render_finished_triangle[next_frame].get_raw_object()},
+                VK_NULL_HANDLE
+            )
+        );
+
+        // RP2
+        app.device()->vkAcquireNextImageKHR(
+            app.device(),
+            app.swapchain().get_raw_object(),
+            UINT64_MAX,
+            image_acquired[current_frame].get_raw_object(),
+            static_cast<VkFence>(VK_NULL_HANDLE),
+            &image_index
+        );
+
+        if (progress_post.find(image_index) != progress_post.end()) {
```

Comment:
Discussed offline, but this is not necessary (Also I think this is causing a hang)
There are 2 waits for any given fence, but only one signal.

Reply:
I removed it and the hang went away. It must have been progress_post.